fdbserver/DDTenantCache.actor.cpp

```diff
@@ -0,0 +1,304 @@
+		TraceEvent(SevInfo, "StartingDDTenantCacheMonitor", tenantCache->id()).log();
+
+		state Transaction tr(tenantCache->dbcx());
+
+		state double lastTenantListFetchTime = now();
+
+		loop {
+			try {
+				if (now() - lastTenantListFetchTime > (2 * SERVER_KNOBS->DD_TENANT_LIST_REFRESH_INTERVAL)) {
+					TraceEvent(SevWarn, "DDTenantListRefreshDelay", tenantCache->id()).log();
+				}
+
+				state RangeResult tenantList = wait(getTenantList(tenantCache, &tr));
+
+				tenantCache->startRefresh();
+				bool tenantListUpdated = false;
+
+				for (int i = 0; i < tenantList.size(); i++) {
+					TenantName tname = tenantList[i].key.removePrefix(tenantMapPrefix);
+					TenantMapEntry t = decodeTenantEntry(tenantList[i].value);
+
+					if (tenantCache->update(tname, t)) {
+						tenantListUpdated = true;
+					}
+				}
+
+				if (tenantCache->cleanup()) {
+					tenantListUpdated = true;
+				}
+
+				if (tenantListUpdated) {
+					TraceEvent(SevInfo, "DDTenantCache", tenantCache->id()).detail("List", tenantCache->desc());
+				}
+
+				lastTenantListFetchTime = now();
+				tr.reset();
+				wait(delay(SERVER_KNOBS->DD_TENANT_LIST_REFRESH_INTERVAL));
+			} catch (Error& e) {
+				if (e.code() != error_code_actor_cancelled) {
+					TraceEvent("DDTenantCacheGetTenantListError", tenantCache->id())
+					    .errorUnsuppressed(e)
+					    .suppressFor(1.0);
+				}
+				wait(tr.onError(e));
+			}
+		}
+	}
+};
+
+void DDTenantCache::insert(TenantName& tenantName, TenantMapEntry& tenant) {
+	KeyRef tenantPrefix(tenant.prefix.begin(), tenant.prefix.size());
+	ASSERT(tenantCache.find(tenantPrefix) == tenantCache.end());
+
+	TenantInfo tenantInfo(tenantName, tenant.id);
+	tenantCache[tenantPrefix] = makeReference<TCTenantInfo>(tenantInfo, tenant.prefix);
+	tenantCache[tenantPrefix]->updateCacheGeneration(generation);
+}
+
+void DDTenantCache::startRefresh() {
+	ASSERT(generation < std::numeric_limits<uint64_t>::max());
+	generation++;
+}
+
+void DDTenantCache::keep(TenantName& tenantName, TenantMapEntry& tenant) {
+	KeyRef tenantPrefix(tenant.prefix.begin(), tenant.prefix.size());
+
+	ASSERT(tenantCache.find(tenantPrefix) != tenantCache.end());
+	tenantCache[tenantPrefix]->updateCacheGeneration(generation);
+}
+
+bool DDTenantCache::update(TenantName& tenantName, TenantMapEntry& tenant) {
+	KeyRef tenantPrefix(tenant.prefix.begin(), tenant.prefix.size());
+
+	if (tenantCache.find(tenantPrefix) != tenantCache.end()) {
+		keep(tenantName, tenant);
+		return false;
+	}
+
+	insert(tenantName, tenant);
+	return true;
+}
+
+int DDTenantCache::cleanup() {
+	int tenantsRemoved = 0;
+	std::vector<Key> keysToErase;
+
+	for (auto& t : tenantCache) {
+		ASSERT(t.value->cacheGeneration() <= generation);
+		if (t.value->cacheGeneration() != generation) {
+			keysToErase.push_back(t.key);
+		}
+	}
+
+	for (auto& k : keysToErase) {
+		tenantCache.erase(k);
+		ASSERT(tenantCache.find(k) == tenantCache.end());
+		tenantsRemoved++;
```

Comment:
I'm not sure whether we should pay the cost of extra `find()` here. Maybe `ASSERT_WE_THINK` is better?

Reply:
Actually, upon revisiting this, I think this is unnecessary and can probably be removed entirely.